XXX-asyncwrap-api.md

```diff
@@ -481,22 +502,92 @@ each bach request execute inside the correct parent.
 This cannot be enforced in any way. It is up to the implementer to make sure
 all of the callbacks are placed and called at the correct time.
 
-```js
-// Here require the "handle" object is passed so it can be maintained in the
-// Map of handles. The return value is the new id of the async event.
-const id = async_wrap.emitInit(handle[, parent_id]);
 
-// If the "parent" is different for this call than the current parent id then
-// pass in a new parent_id.
-async_wrap.emitBefore(id[, parent_id]);
+#### `async_wrap.createEvent(handle[, parent_id])
+
+* `handle` {Object}
+* `parent_id` {Number} **Default:** `async_wrap.currentId()`
+* Return: {AsyncEvent}
+
+Returns an instance of `AsyncEvent` to be used to trigger `AsyncHook` callback.
+
+`handle` is the object instance that's passed to `init()`.
+
+`parent_id` is the value passed as `parentId` to `init()`. If omitted then the
+current async id is used.
+
+An instance of `AsyncEvent` is returned that is to be used to trigger the
+`AsyncHook` callbacks.
+
+
+#### `event.getId()`
+
+* Return: {Number}
+
+Returns the unique id assigned to the instance when it was created.
+
+Would be easier to just slap on the number, but right now I'm not 100% sure
+that the implementation won't change in the future. So sticking with this.
+
+
+#### `event.emitInit()`
 
-async_wrap.emitAfter(id[, parent_id]);
+* Return: {Undefined}
 
-// The "handle" in the Map will be removed on emitDestroy(). If this is not
-// called then it will lead to a memory leak.
-async_wrap.emitDestroy(id[, parent_id]);
+Trigger listening `init()` callbacks that this event has been created.  Will
+throw if `emitInit()` is called twice.
+
+
+#### `event.emitBefore()`
+
+* Return: {Undefined}
+
+Trigger listening `before()` callbacks that the `handle` is about to enter it's
+call stack.
+
+
+#### `event.emitAfter()`
+
+* Return: {AsyncEvent}
+
+Trigger listening `after()` callbacks that the `handle` has left it's call
+stack.
+
+The `id` call stack should always properly unwind. No two `emitAfter()` calls
+should be crossed. For example:
+
+```
+init # Foo
+init # Bar
+...
+before # Foo
+before # Bar
+after # Bar   <- Should be called after Foo
+after # Foo
 ```
 
+There is no known scenario where this is acceptable. If one is found then this
+restriction may be lifted.
+
+Return value is the calling `AsyncEvent` instance. This is done to make the
```

Comment:
Either the `AsyncEvent` instance should be returned from all the methods or they should all return `undefined`. This is inconsistent.

Personally I would prefer `undefined` being returned, but as long as it is consistent it is fine.